### PR TITLE
fri_cosets now correctly processes arbitrary input

### PR DESCRIPTION
**Pull request by ayashunsky**

### CODE REVIEW

File: include/nil/blueprint/components/systems/snark/plonk/placeholder/fri_cosets.hpp [hunk 1]
```diff
@@ -332,107 +367,142 @@ namespace nil {
...
+                        W0next = var(component.W(9*((j+1) % l)),(j+1)/l),
+                        W1next = var(component.W(9*((j+1) % l) + 1),(j+1)/l),
+                        W2next = var(component.W(9*((j+1) % l) + 2),(j+1)/l),
+                        W3prev = var(component.W(9*((l+j-1) % l) + 3), -(j == 0)),
+                        W4prev = var(component.W(9*((l+j-1) % l) + 4), -(j == 0)),
+                        W5next = var(component.W(9*((j+1) % l) + 5),(j+1)/l);
+
+                    nine_block[9*j]   = W0 - 2*W0next - W7;
+                    nine_block[9*j+1] = W1 - 2*W1next - W8;
+                    nine_block[9*j+2] = (W8 - W7)*(1 - W2next)*(1 + W2next) + (W2next * W2next * W2next) - W2;
```
Comment (Iluvmagick) [inline include/nil/blueprint/components/systems/snark/plonk/placeholder/fri_cosets.hpp:404]: Why is the cubing required here? It seems to me that just doing `+ W2next` should suffice.

File: include/nil/blueprint/components/systems/snark/plonk/placeholder/fri_cosets.hpp [hunk 2]
```diff
@@ -332,107 +367,142 @@ namespace nil {
...
+                        mid1 *= 2;
+                        mid1 += var(component.W(j+1),0);
                     }
-                    mid -= var(component.W(6*last_l),0);
+                    mid  -= var(component.W(9*last_l),0);
+                    mid1 -= var(component.W(9*last_l + 1),0);
                     cs3.push_back(mid);
+                    cs3.push_back(mid1);
+
+                    // the sign bit shoudl be just a copy from the next block
```
Comment (Iluvmagick) [inline include/nil/blueprint/components/systems/snark/plonk/placeholder/fri_cosets.hpp:486]: typo: shou**dl**

File: include/nil/blueprint/components/systems/snark/plonk/placeholder/fri_cosets.hpp [hunk 3]
```diff
@@ -332,107 +367,142 @@ namespace nil {
...
+                    var Wl2     = var(component.W(9*last_l + 2),0),
+                        Wl2next = var(component.W(9*last_l + 5),0);
+                    cs3.push_back( Wl2 - Wl2next );
```
Comment (Iluvmagick) [inline include/nil/blueprint/components/systems/snark/plonk/placeholder/fri_cosets.hpp:489]: Given that we have to assert this equality in only a single row a copy constraint should be more efficient? Unsure.

**Review by Iluvmagick — APPROVED** [review]
Looks good!

File: include/nil/blueprint/components/systems/snark/plonk/placeholder/fri_cosets.hpp (hunk 2 above)
Comment (ayashunsky) [reply include/nil/blueprint/components/systems/snark/plonk/placeholder/fri_cosets.hpp:486]: Fixed

File: include/nil/blueprint/components/systems/snark/plonk/placeholder/fri_cosets.hpp (hunk 1 above)
Comment (ayashunsky) [reply include/nil/blueprint/components/systems/snark/plonk/placeholder/fri_cosets.hpp:404]: That's right. No idea, why I wanted to cube so much.

File: include/nil/blueprint/components/systems/snark/plonk/placeholder/fri_cosets.hpp [hunk 4]
```diff
@@ -332,107 +367,142 @@ namespace nil {
...
-                        W2prev = var(component.W(6*((l+j-1) % l) + 2), -(j == 0)),
-                        W3next = var(component.W(6*((j+1) % l) + 3),(j+1)/l);
-
-                    six_block[6*j]   = W0 - 2*W0next - W5;
-                    six_block[6*j+1] = W1 - W1prev * W1prev;
-                    six_block[6*j+2] = W2 - W2prev * (W1*W5 + 1-W5);
-                    six_block[6*j+3] = W3 - W3next * W3next;
-                    six_block[6*j+4] = W3 + W4;
-                    six_block[6*j+5] = (1 - W5) * W5;
+                    var W0 = var(component.W(9*j),0),
```
[the remark this answers is not in the document]
Comment (ayashunsky) [reply include/nil/blueprint/components/systems/snark/plonk/placeholder/fri_cosets.hpp:386]: Added comments to explain the column usage. I'd rather leave the column names as they are now and stick to descriptive names in future components.

File: include/nil/blueprint/components/systems/snark/plonk/placeholder/fri_cosets.hpp (hunk 3 above)
Comment (ayashunsky) [reply include/nil/blueprint/components/systems/snark/plonk/placeholder/fri_cosets.hpp:489]: It seems to me, that moving this to copy constraints slightly disrupts the logic. We would then have a 3-block with some parts encoded as polynomial constraints and other parts as copy constraits, whereas other 3-blocks are all encoded with just poly-constraints. I would move this to copy constraints only if there is a clear performance advantage.

**Pull request merged**